Sources/ComposedUI/CollectionView/CollectionCoordinator.swift

```diff
@@ -163,8 +165,13 @@ open class CollectionCoordinator: NSObject {
 
             switch section.cell.dequeueMethod {
             case let .fromNib(type):
-                let nib = UINib(nibName: String(describing: type), bundle: Bundle(for: type))
+                let nibName = String(describing: type)
+                guard !registeredNibNames.contains(nibName) else { break }
+                // For some reason, large amount of call to `UINib(nibName:bundle:)` will significantly impact performance when App just
+                // started up, but fine later on. Avoid calling if nib has already been registered
```

Comment:
Interesting that it's only slow the first time it's loaded. I'm guessing the unarchived nib is kept in memory but I'm interested in learning more.

If this is primarily an issue when the app first starts/a screen first loads it might also work to register the cell immediately before dequeuing it so a cell is not registered if it starts off screen?

Reply:
You can indeed  register before dequeuing, but in terms of code structure, it looks better at the beginning.

I am wondering why there is a slow down in this case too, but unlike Android, iOS's UIKit is not open source? So it might be difficult to find out the actual reason for the slow down.

There are too many tasks at hand, perhaps when we have time, we can look for a proper solution, or perhaps Shaps80 can help and look into it after the holidays?

Reply:
> You can indeed register before dequeuing, but in terms of code structure, it looks better at the beginning.

There are a few pieces of Composed that aren't the clearest or most concise code but lead to a better API and improved performance. Heavily documenting these parts of the code usually counters the overall poorer code.

> I am wondering why there is a slow down in this case too, but unlike Android, iOS's UIKit is not open source? So it might be difficult to find out the actual reason for the slow down.

Some parts are (https://opensource.apple.com/) but the parts we're dealing with (and the vast majority of anything UI related) isn't.

I'm pretty sure the performance hit of the first load is due to loading the nibs from disk. [The documentation](https://developer.apple.com/documentation/uikit/uinib) shows that it stores the contents in memory (and even frees that memory in low memory conditions), but it doesn't quite describe what we're seeing: subsequent calls to `UINib.init(nibName:bundle:)` with the same parameters are much faster. I would assume they have an internal cache that speeds this up.

> There are too many tasks at hand, perhaps when we have time, we can look for a proper solution, or perhaps Shaps80 can help and look into it after the holidays?

Yep, each change like this we can either create PRs for in to the main repo or create issues for the ones we're less sure about.

Reply:
Okay, good to know